### PR TITLE
Update event log structure

### PR DESCRIPTION
**Pull request by dadlerj**
I've tagged a few people who I think would be good subject matter experts for parts of this!

* @camdencheek you gave me some great advice on refactoring the `siteid` package in Slack (https://sourcegraph.slack.com/archives/C042NF2ECA1/p1689368401080439). I've moved all of those changes to [a single logical commit](https://github.com/sourcegraph/sourcegraph/commit/af8a4e991edb4e7e05bb1849304e7bb234a4c262) so you can review easily
* @sourcegraph/security  you all gave me some good advice in slack on adding the license key hashing function. I've tagged you here as suggested by @willdollman. I've grouped these changes into [a single logical commit](https://github.com/sourcegraph/sourcegraph/commit/136683e262d2d485d794a283932c891530964d67) so you can review easily.
* @coury-clark you have a lot of background with our overall telemetry/eventlogger framework and I'd love your input on the core changes (probably the first commit to add to the database and the last one to update the event structure). If you think someone else would be better for this, please let me know.
* Obviously @nathan-downs and @akalia25 have been instrumental in building this plan, so I'd love any input on the overall update and the last commit in particular.

## PR overview

Update event logs schema to include:

* Connected site identification values (`connected_site_id`, `hashed_license_key`). With the focus on IDE extensions, we lack the ability to link IDE usage to servers and customers. This will enable that by adding two anonymous identifiers to all event logging clients that can be used to make that link.
* A client value. Similarly, today we lack a single property that distinguishes the client being used, and we rely on a mix of source and name to build this.
* Two billing values. Not yet ready to start using given a large migration of existing event calls will be required, but these will be necessary when we begin to license code search and cody separately.

This PR is only step https://github.com/sourcegraph/sourcegraph/issues/1. Once merged, we must add these new properties to every eventlogger client.

Other changes:

* Add a new GQL endpoint to get a hashed license key. This will be used by other Sourcegraph clients (e.g. IDE extensions) to get the value to send with event logs
* Move and refactor the `siteid` package. Necessary to access the site ID from other /internal packages

## Test plan

* Test locally to ensure event properties are being added as expected.
* Check BigQuery to ensure event properties are being successfully passed.

### CODE REVIEW

File: migrations/frontend/1689176578_add_new_fields_to_event_logs_table/up.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE IF EXISTS event_logs
```
Comment (coury-clark) [inline migrations/frontend/1689176578_add_new_fields_to_event_logs_table/up.sql:1]: There is a comment elsewhere that we aren't intending on logging these new fields to the DB, so I'm not clear on whether that is true, or if these fields are necessary? Could you clarify?

File: internal/usagestats/event_handlers.go [hunk 2]
```diff
@@ -40,24 +42,45 @@ type Event struct {
...
+	PublicArgument         json.RawMessage
+	UserProperties         json.RawMessage
+	DeviceID               *string
+	InsertID               *string
+	EventID                *int32
+	DeviceSessionID        *string
+	Client                 *string
+	BillingProductCategory *string
+	BillingEventID         *string
+	// ConnectedSiteID is only logged for Cloud events; therefore, this only goes to the BigQuery database
```
[the remark this answers is not in the document]
Comment (dadlerj) [reply internal/usagestats/event_handlers.go:59]: Thanks for the context, that's interesting...

There are two different code paths in this file:

1. First, [events that get logged into the local database's `event_logs` table](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4049ecb98cbfa142f583c073d1e4105a3c4121e1/-/blob/internal/usagestats/event_handlers.go?L217:6-217:20). 
2. Second, [events that get sent to GCP pub/sub and BigQuery](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/usagestats/event_handlers.go?L127:6-127:36#tab=references) (our data warehouse)

In general, the comments here reflect the fact that fields like `connected_site_id` aren't important for the first category. If an event is logged in a local postgres db, then the connected site is ALWAYS that server that hosts that postgres db. That's what this comment is referring to. Other fields, like `client` ARE still important for a local db. 

I'm not entirely sure how the cloud (read: managed instance) data pipeline works, but if it's just scraping the cloud instance DBs, then I agree it could be problematic to not include these fields there (since they would eventually get scraped and aggregated with other instance data and the assumption that the `connected_site_id` is trivial would be lost). 

I probably need to understand more about how our single tenant cloud instance event logging works... If you have that context, let me know!

File: migrations/frontend/1689176578_add_new_fields_to_event_logs_table/up.sql (hunk 1 above)
Comment (dadlerj) [reply migrations/frontend/1689176578_add_new_fields_to_event_logs_table/up.sql:1]: These three fields ARE important for the local DB, while `connected_site_id` and `hashed_license_key` are not (since on a local instance, they would always just reflect that instance's site_id and license key.

File: internal/usagestats/event_handlers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nathan-downs) [reply internal/usagestats/event_handlers.go:59]: We have a couple different paths currently that I tried to capture in this diagram: https://www.figma.com/file/H8ipJVvKEWbx5TqnGDsjXU/Event-Logging-Everywhere-Architecture?type=whiteboard&node-id=0-1&t=x076FpNfT8L3CtzW-0

[the remark this answers is not in the document]
Comment (coury-clark) [reply internal/usagestats/event_handlers.go:59]: @dadlerj You can also find the [original RFC](https://docs.google.com/document/d/1N9aO0uTlvwXI7FzdPjIUCn_d1tRkJUfsc0urWigRf6s/edit) when I implemented the managed instance pipeline, there is a diagram similar to the one linked above.

tl;dr - eventLogger -> frontend -> postgres -> background job scrapes them -> bigquery

Here is the [event struct we send to BigQuery for managed instances](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/worker/internal/telemetry/telemetry_job.go?L308-330)

[the remark this answers is not in the document]
Comment (dadlerj) [reply internal/usagestats/event_handlers.go:59]: Amazing, thank you. @coury-clark that file is what I was hoping to see. I think the comment in the file above is still correct. We don't need to log `connected_site_id` or `hashed_license_key` in a server postgres table. Instead, I can just add them to the struct above and add lines like:

```
import "github.com/sourcegraph/sourcegraph/internal/conf"
import "github.com/sourcegraph/sourcegraph/internal/siteid"

return &bigQueryEvent{
   ...
   ConnectedSiteId: siteid.Get(),
   HashedLicenseKey: conf.HashedCurrentLicenseKeyForAnalytics(),
...

```

Since, EVERY event on, say, s2 will have the same connected site id and license key, right?

[the remark this answers is not in the document]
Comment (coury-clark) [reply internal/usagestats/event_handlers.go:59]: At a high level that sounds right, each instance should have the same ID. I think `connected_site_id` came after my time here, so I'm not _entirely_ sure if that's different from site_id, but conceptually this makes sense.

As far as the code update goes, we load this instance metadata object that gets populated for each event (rather than calling this once per event since it's a big batch). You can add the load [here](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/worker/internal/telemetry/telemetry_job.go?L394-412), and then just pass it from this metadata struct into the event struct when the event is being formed (link above).

[the remark this answers is not in the document]
Comment (dadlerj) [reply internal/usagestats/event_handlers.go:59]: Yeah—good question. `connected_site_id` is just the site id of the connected Sourcegraph instance to any event logged. E.g., if I'm using the VS Code Cody extension, and I'm signed in to s2, a recipe or completion event would be logged to s2's `event_logs` table with a `connected_site_id` of s2's site_id. 

And if an event was logged on s2's web app itself, or from s2's backend, s2's site ID would be used. 

And thanks for the details. I'll resolve this conversation unless you have any concerns.

**Review by evict — APPROVED** [review]
Looks good from a security PoV. Adding the prefix to the license key before hashing is a good way to make the hashing more effective. 👍
